In [1]:
import numpy as np
import pandas as pd

In [2]:
df_ratings = pd.read_csv('/content/ratings.csv',header = None, sep = '::',names = ["UserId","MovieId","Rating","Timestamp"])
df_ratings

<ipython-input-2-871468e1d80f>:1: ParserWarning: Falling back to the 'python' engine because the 'c' engine does not support regex separators (separators > 1 char and different from '\s+' are interpreted as regex); you can avoid this warning by specifying engine='python'.
  df_ratings = pd.read_csv('/content/ratings.csv',header = None, sep = '::',names = ["UserId","MovieId","Rating","Timestamp"])


UserId  MovieId  Rating  Timestamp
0             1     1193       5  978300760
1             1      661       3  978302109
2             1      914       3  978301968
3             1     3408       4  978300275
4             1     2355       5  978824291
...         ...      ...     ...        ...
1000204    6040     1091       1  956716541
1000205    6040     1094       5  956704887
1000206    6040      562       5  956704746
1000207    6040     1096       4  956715648
1000208    6040     1097       4  956715569

[1000209 rows x 4 columns]

In [3]:
df_user = pd.read_csv('/content/users.csv',header = None, sep = '::', names = ["UserId","Gender","Age","Occupation","ZipCode"])
df_user

<ipython-input-3-50c72de4bf3b>:1: ParserWarning: Falling back to the 'python' engine because the 'c' engine does not support regex separators (separators > 1 char and different from '\s+' are interpreted as regex); you can avoid this warning by specifying engine='python'.
  df_user = pd.read_csv('/content/users.csv',header = None, sep = '::', names = ["UserId","Gender","Age","Occupation","ZipCode"])


UserId Gender  Age  Occupation ZipCode
0          1      F    1          10   48067
1          2      M   56          16   70072
2          3      M   25          15   55117
3          4      M   45           7   02460
4          5      M   25          20   55455
...      ...    ...  ...         ...     ...
6035    6036      F   25          15   32603
6036    6037      F   45           1   76006
6037    6038      F   56           1   14706
6038    6039      F   45           0   01060
6039    6040      M   25           6   11106

[6040 rows x 5 columns]

In [4]:
df_movies = pd.read_csv('/content/movies.csv',encoding='latin-1',on_bad_lines='skip',header=None, names= ["MovieId","Title","Genres"],sep= '::')
df_movies

<ipython-input-4-39e10dc8cfda>:1: ParserWarning: Falling back to the 'python' engine because the 'c' engine does not support regex separators (separators > 1 char and different from '\s+' are interpreted as regex); you can avoid this warning by specifying engine='python'.
  df_movies = pd.read_csv('/content/movies.csv',encoding='latin-1',on_bad_lines='skip',header=None, names= ["MovieId","Title","Genres"],sep= '::')


MovieId                               Title  \
0           1                    Toy Story (1995)   
1           2                      Jumanji (1995)   
2           3             Grumpier Old Men (1995)   
3           4            Waiting to Exhale (1995)   
4           5  Father of the Bride Part II (1995)   
...       ...                                 ...   
3878     3948             Meet the Parents (2000)   
3879     3949          Requiem for a Dream (2000)   
3880     3950                    Tigerland (2000)   
3881     3951             Two Family House (2000)   
3882     3952               Contender, The (2000)   

                            Genres  
0      Animation|Children's|Comedy  
1     Adventure|Children's|Fantasy  
2                   Comedy|Romance  
3                     Comedy|Drama  
4                           Comedy  
...                            ...  
3878                        Comedy  
3879                         Drama  
3880                         Drama  
3881                         Drama  
3882                Drama|Thriller  

[3883 rows x 3 columns]

Creating Pivot Table for gaining more insights and to apply Collaborative Learning

In [5]:
pivot_table = df_ratings.pivot_table(values='Rating',columns='UserId',index='MovieId').fillna(0)

pivot_table.head()

UserId   1     2     3     4     5     6     7     8     9     10    ...  \
MovieId                                                              ...   
1         5.0   0.0   0.0   0.0   0.0   4.0   0.0   4.0   5.0   5.0  ...   
2         0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   5.0  ...   
3         0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   
4         0.0   0.0   0.0   0.0   0.0   0.0   0.0   3.0   0.0   0.0  ...   
5         0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   

UserId   6031  6032  6033  6034  6035  6036  6037  6038  6039  6040  
MovieId                                                              
1         0.0   4.0   0.0   0.0   4.0   0.0   0.0   0.0   0.0   3.0  
2         0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
3         0.0   0.0   0.0   0.0   1.0   0.0   0.0   0.0   0.0   0.0  
4         0.0   0.0   0.0   0.0   2.0   2.0   0.0   0.0   0.0   0.0  
5         0.0   0.0   0.0   0.0   1.0   0.0   0.0   0.0   0.0   0.0  

[5 rows x 6040 columns]

In [6]:
print(pivot_table.shape)

(3706, 6040)


In [7]:
from sklearn.model_selection import train_test_split

train_data, test_data = train_test_split(pivot_table, test_size=0.2, random_state=2023)

In [8]:
train_data

UserId   1     2     3     4     5     6     7     8     9     10    ...  \
MovieId                                                              ...   
1197      3.0   0.0   5.0   0.0   0.0   0.0   0.0   0.0   0.0   5.0  ...   
1012      0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   3.0  ...   
1322      0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   
1934      0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   
741       0.0   0.0   0.0   0.0   0.0   0.0   0.0   5.0   0.0   0.0  ...   
...       ...   ...   ...   ...   ...   ...   ...   ...   ...   ...  ...   
2134      0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   
2951      0.0   0.0   0.0   4.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   
2707      0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   
1702      0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   
916       0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   5.0  ...   

UserId   6031  6032  6033  6034  6035  6036  6037  6038  6039  6040  
MovieId                                                              
1197      0.0   3.0   0.0   0.0   0.0   5.0   0.0   0.0   4.0   0.0  
1012      0.0   0.0   4.0   0.0   0.0   4.0   0.0   0.0   0.0   0.0  
1322      0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
1934      0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   5.0   0.0  
741       0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
...       ...   ...   ...   ...   ...   ...   ...   ...   ...   ...  
2134      0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   1.0  
2951      0.0   0.0   3.0   0.0   0.0   4.0   0.0   0.0   0.0   0.0  
2707      0.0   0.0   0.0   0.0   0.0   2.0   0.0   0.0   0.0   0.0  
1702      0.0   0.0   0.0   0.0   2.0   0.0   0.0   0.0   0.0   0.0  
916       0.0   0.0   0.0   0.0   0.0   5.0   3.0   0.0   5.0   5.0  

[2964 rows x 6040 columns]

In [9]:
test_data

UserId   1     2     3     4     5     6     7     8     9     10    ...  \
MovieId                                                              ...   
2811      0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   
306       0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   
2870      0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   
166       0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   
2751      0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   
...       ...   ...   ...   ...   ...   ...   ...   ...   ...   ...  ...   
3781      0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   
690       0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   
2941      0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   
2019      0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   
3363      0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   5.0  ...   

UserId   6031  6032  6033  6034  6035  6036  6037  6038  6039  6040  
MovieId                                                              
2811      0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
306       0.0   0.0   0.0   0.0   5.0   0.0   0.0   0.0   0.0   4.0  
2870      0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
166       0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
2751      0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   1.0  
...       ...   ...   ...   ...   ...   ...   ...   ...   ...   ...  
3781      0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
690       0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
2941      0.0   0.0   0.0   0.0   0.0   3.0   0.0   0.0   0.0   0.0  
2019      0.0   0.0   0.0   0.0   0.0   5.0   0.0   0.0   4.0   5.0  
3363      0.0   4.0   0.0   0.0   0.0   2.0   5.0   0.0   5.0   0.0  

[742 rows x 6040 columns]

For collaborative filtering, models like matrix factorization, nearest neighbors, or deep learning methods are more common and effective.

I will be using Nearest Neighbors model in this code.

In [10]:
from sklearn.neighbors import NearestNeighbors

model = NearestNeighbors(metric='cosine')
model.fit(train_data)

NearestNeighbors(metric='cosine')

In [11]:
# Function to predict ratings for the test set
def predict_ratings(user_index):
    distances, indices = model.kneighbors(test_data.iloc[user_index, :].values.reshape(1, -1), n_neighbors=5)
    neighbor_ratings = train_data.iloc[indices[0], :]
    predicted_ratings = np.mean(neighbor_ratings, axis=0)
    return predicted_ratings

In [12]:
predicted_ratings = [predict_ratings(user_index) for user_index in range(len(test_data))]

In [13]:
actual_ratings_flat = test_data.values.flatten()
predicted_ratings_flat = np.concatenate(predicted_ratings)

In [17]:
from sklearn.metrics import mean_squared_error
from sklearn.metrics import confusion_matrix
import math

In [18]:
rmse = math.sqrt(mean_squared_error(actual_ratings_flat, predicted_ratings_flat))
print(f"Root Mean Squared Error (RMSE): {rmse}")

Root Mean Squared Error (RMSE): 0.7107228819092742


This value indicates that, on average, our model's predictions are off by approximately 0.7107 units on a scale of 1 to 5.

Lower RMSE values indicates better performance of the model.
So getting the RMSE value equal to 0.7107 indicates better performance of Nearest Neighbours model.

In [21]:
# Converting ratings to binary for confusion matrix
# (If rating >= 3, considering it as positive)
actual_binary = (actual_ratings_flat >= 3).astype(int)
predicted_binary = (predicted_ratings_flat >= 3).astype(int)


In [22]:
conf_matrix = confusion_matrix(actual_binary, predicted_binary)
print("Confusion Matrix:")
print(conf_matrix)

Confusion Matrix:
[[4277294   38625]
 [ 108669   57092]]


In [23]:
# So let's traing the model on whole dataset
model.fit(pivot_table)

NearestNeighbors(metric='cosine')

In [24]:
class Recommender:
    def __init__(self):
        # To stroe the list of recommendation movies history.
        self.hist = []
        self.ishist = False # For checking whether history is empty or not.

    # This method will recommend movies based on a movie that is passed as the parameter.
    # This is for those times when you know you have watched a particular movie and want to watch similar movies.
    # I am setting the function to recommend 5 movies
    def recommend_on_movie(self,movie,n_reccomend = 5):
        self.ishist = True
        movie_id = int(df_movies[df_movies['Title']==movie]['MovieId'])
        self.hist.append(movie_id) #Adding the movie id in list

        distance,neighbors = model.kneighbors([pivot_table.loc[movie_id]],n_neighbors=n_reccomend+1)
        movie_ids = [pivot_table.iloc[i].name for i in neighbors[0]]
        recommendations = [str(df_movies[df_movies['MovieId']==mid]['Title']).split('\n')[0].split('  ')[-1] for mid in movie_ids if mid not in [movie_id]]
        return recommendations[:n_reccomend]

    # This method will recommend movies based on history stored in self.hist list
    # This is for those times when you know you have watched movies and based on all those movies, you want recommendation.
    # I am setting the function to recommend 5 movies
    def recommend_on_history(self,n_reccomend = 5):
        if self.ishist == False:
            return print('No history found')
        history = np.array([list(pivot_table.loc[mid]) for mid in self.hist])
        distance,neighbors = model.kneighbors([np.average(history,axis=0)],n_neighbors=n_reccomend + len(self.hist))
        movie_ids = [pivot_table.iloc[i].name for i in neighbors[0]]
        recommendations = [str(df_movies[df_movies['MovieId']==mid]['Title']).split('\n')[0].split('  ')[-1] for mid in movie_ids if mid not in self.hist]
        return recommendations[:n_reccomend]

In [25]:
# Initializing the Recommender Object
recommender = Recommender()

In [26]:
recommender.recommend_on_history()

No history found


Genre of Jumanji is Adventure|Children's|Fantasy

In [27]:
recommender.recommend_on_movie('Jumanji (1995)')

['Hook (1991)',
 'Dragonheart (1996)',
 'Indian in the Cupboard, The (1995)',
 'Honey, I Shrunk the Kids (1989)',
 'NeverEnding Story, The (1984)']

Let's check the genres of movies recommended by our model and see whether the genre are similar or not.

In [39]:
df_movies.loc[df_movies['Title']=='Hook (1991)']

MovieId        Title             Genres
3420     3489  Hook (1991)  Adventure|Fantasy

In [40]:
df_movies.loc[df_movies['Title']=='Indian in the Cupboard, The (1995)']

MovieId                               Title                        Genres
59       60  Indian in the Cupboard, The (1995)  Adventure|Children's|Fantasy

In [41]:
df_movies.loc[df_movies['Title']=='Dragonheart (1996)']

MovieId               Title                    Genres
647      653  Dragonheart (1996)  Action|Adventure|Fantasy

In [43]:
df_movies.loc[df_movies['Title']=='Honey, I Shrunk the Kids (1989)']

MovieId                            Title  \
1985     2054  Honey, I Shrunk the Kids (1989)   

                                          Genres  
1985  Adventure|Children's|Comedy|Fantasy|Sci-Fi

In [42]:
df_movies.loc[df_movies['Title']=='NeverEnding Story, The (1984)']

MovieId                          Title                        Genres
2092     2161  NeverEnding Story, The (1984)  Adventure|Children's|Fantasy

As you can see they have same genre so our model recommendation is good based on genre matching.

Let's also check the ratings of these movies.

In [49]:
# first checking the rating of the movie Jumanji (1995)

movie = df_ratings.loc[df_ratings['MovieId']==2]

movie

UserId  MovieId  Rating  Timestamp
839         10        2       5  979168267
1367        13        2       3  978202563
1967        18        2       2  978152541
2871        23        2       2  978461604
3778        27        2       1  978129692
...        ...      ...     ...        ...
993046    6000        2       3  956884937
993544    6001        2       3  956805460
994516    6006        2       2  957032380
996986    6019        2       4  956761170
998006    6030        2       4  956719238

[701 rows x 4 columns]

So there are 701 users who have given the ratings to this movie.

The ratings will differ based on the user's way of thinking as they are of different age and have different occupation. Also, gender plays a role too

If we take the mean of the ratings here from all users

In [50]:
print(movie['Rating'].mean())

3.20114122681883


In [51]:
# Indian in the Cupboard, The (1995) has movieid = 60

movie = df_ratings.loc[df_ratings['MovieId']==60]

print(movie['Rating'].mean())

3.212885154061625


We can see, it has similar mean rating from the user.

Let's check for one another movie too.

In [52]:
# NeverEnding Story, The (1984) has movieid = 2161

movie = df_ratings.loc[df_ratings['MovieId']==2161]

print(movie['Rating'].mean())

3.6051502145922747


In [53]:
# Honey, I Shrunk the Kids (1989) has movieid = 2054

movie = df_ratings.loc[df_ratings['MovieId']==2054]

print(movie['Rating'].mean())

2.9330143540669855


Here we can see that we are getting the movies of similar mean ratings.

Although as I told, mean value doesn't hold much importance because different users have different aspects which are guided by gender, age, occupation and place where they live.

In [28]:
recommender.recommend_on_history()

['Hook (1991)',
 'Dragonheart (1996)',
 'Indian in the Cupboard, The (1995)',
 'Honey, I Shrunk the Kids (1989)',
 'NeverEnding Story, The (1984)']

In [29]:
recommender.recommend_on_movie('Meet the Parents (2000)')

['Almost Famous (2000)',
 'Best in Show (2000)',
 'Nurse Betty (2000)',
 'High Fidelity (2000)',
 'Remember the Titans (2000)']

In [30]:
recommender.recommend_on_history()

['Jurassic Park (1993)',
 'Men in Black (1997)',
 'Almost Famous (2000)',
 'Hook (1991)',
 'Mission: Impossible (1996)']

In [31]:
recommender.recommend_on_movie('Toy Story (1995)')

['Toy Story 2 (1999)',
 'Groundhog Day (1993)',
 'Aladdin (1992)',
 "Bug's Life, A (1998)",
 'Back to the Future (1985)']

In [32]:
recommender.recommend_on_history()

['Groundhog Day (1993)',
 'Aladdin (1992)',
 'Toy Story 2 (1999)',
 'Men in Black (1997)',
 'Matrix, The (1999)']